external/kafka/src/main/scala/org/apache/spark/rdd/kafka/OffsetRange.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd.kafka
+
+/** Represents a range of offsets from a single Kafka TopicAndPartition */
+trait OffsetRange {
```

Comment:
How about something like this. 
```
class TopicPartitionOffsetRange private[kafka] (
   val topic: String, 
   val partition: .....
   val fromOffset: ...
   val untilOffset: ...
) {
   

}
```
and then we can define KafkaRDDPartition as 

```
private[kafka] class KafkaRDDPartition(offsetRange: OffsetRange, host: String, port: Int) extends Partition
```

- Does not expost KafkaRDDPartition
- Does not use Scala trait or Java interface (requires separate java file, annoying)
- People can publicly define OffsetRanges

If we want to be able to define sets of OffsetRange (not sure if we need to), we can implement hashCode and equals as well. Case class would have given that for free, but the automatically defined object method that come with case classes can become annoying for Java API compatibility. Better to steer clear.